jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,665 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import org.eclipse.jetty.http.pathmap.PathSpec;
+import org.eclipse.jetty.http.pathmap.PathSpecSet;
+import org.eclipse.jetty.server.ConnectionMetaData;
+import org.eclipse.jetty.server.Connector;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IncludeExclude;
+import org.eclipse.jetty.util.IncludeExcludeSet;
+import org.eclipse.jetty.util.InetAddressPattern;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link Handler.Wrapper} that may conditionally apply an action to a request.
+ * The conditions are implemented by an {@link IncludeExclude} of:
+ * <ul>
+ *     <li>A method name, which can be efficiently matched</li>
+ *     <li>A {@link PathSpec} or sting representation, which can be efficient matched.</li>
+ *     <li>An arbitrary {@link Predicate} taking the {@link Request}, which is matched in a linear test of all predicates.</li>
+ * </ul>
+ * <p>
+ * If the conditions are met, the {@link #doHandle(Request, Response, Callback)} method will be invoked.
+ * However, as the default implementation to to call the {@link #getHandler() next Handler}, an optimization is applied to
+ * directly call the next {@code Handler} if {@code doHandler} has not been extended.
+ * </p>
+ * <p>
+ * If the conditions are not met, then the behaviour will be determined by the {@link ConditionNotMetAction} passed to the
+ * constructor.
+ * </p>
+ *
+ */
+public class ConditionalHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ConditionalHandler.class);
+
+    /**
+     * Enumeration of the actions that can be taken if the request is to not be handled due to failing to
+     * meet the conditions.
+     */
+    public enum ConditionNotMetAction
```

Comment:
`NoMatchAction`?

Reply:
@sbordet  It is shorter and I could live with it.... but I've not used "match" with "condition" any where else in the code/javadoc.     Currently the class is described in terms of matching or notMatching a condition, which results in handling or notHandling the request.   So that is consistent.  Adding in "match" just for a shorter enum name doesn't really add clarity.   Ditto for "apply".